README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# nsapi

[![CRAN\_latest\_release\_date](https://www.r-pkg.org/badges/last-release/nsapi)](https://cran.r-project.org/package=nsapi)
[![license](https://img.shields.io/github/license/mashape/apistatus.svg)](https://choosealicense.com/licenses/mit/)
[![Project Status: Active – The project has reached a stable, usable
state and is being actively
developed.](http://www.repostatus.org/badges/latest/active.svg)](http://www.repostatus.org/#active)
[![thanks-md](https://img.shields.io/badge/THANKS-md-ff69b4.svg)](THANKS.md)
[![lifecycle](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing)
[![Travis-CI Build
Status](https://travis-ci.org/RMHogervorst/nsapi.svg?branch=master)](https://travis-ci.org/RMHogervorst/nsapi)
[![codecov](https://codecov.io/gh/RMHogervorst/nsapi/branch/master/graph/badge.svg)](https://codecov.io/gh/RMHogervorst/nsapi)

The Dutch National Railway service (NS; Nederlandse Spoorwegen) has an
API where we can query for travel advise, see the current trains on a
given station, see if there is any delays or work on the tracks and NS
also provides a list with geolocation of all the stations.

The goal of nsapi is to make it easy to gather data from the NS API. The
package returns data frames for every response.

![an incredibly ugly logo for this package, we need a
hexsticker\!](man/figures/nsapilogo.png)

## Installation

You can NOT YET install the released version of nsapi from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("nsapi")
```

But you can install the development version from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("RMHogervorst/nsapi")
```

## Example

This is a basic example which shows you how you get travel information:

``` r
library(nsapi)
```

``` r
treinplanner <- get_travel_advise(
  fromStation = "Leiden Centraal", 
  toStation = "Utrecht Centraal",
  departure = TRUE,
  yearCard = TRUE,
  previousAdvises = 1, 
  nextAdvises = 1)
treinplanner
#>   Melding AantalOverstappen GeplandeReisTijd ActueleReisTijd
#> 1      NA                 1             0:55            0:58
#> 2      NA                 0             0:42            0:42
#> 3      NA                 1             1:00            1:00
#> 4      NA                 0             0:42            0:42
#>   VertrekVertraging AankomstVertraging Optimaal GeplandeVertrekTijd
#> 1              <NA>             +3 min    FALSE 2018-08-10 21:05:00
#> 2              <NA>               <NA>     TRUE 2018-08-10 21:22:00
#> 3              <NA>               <NA>    FALSE 2018-08-10 21:30:00
#> 4              <NA>               <NA>    FALSE 2018-08-10 21:52:00
#>    ActueleVertrekTijd GeplandeAankomstTijd ActueleAankomstTijd
#> 1 2018-08-10 21:05:00  2018-08-10 22:00:00 2018-08-10 22:03:00
#> 2 2018-08-10 21:22:00  2018-08-10 22:04:00 2018-08-10 22:04:00
#> 3 2018-08-10 21:30:00  2018-08-10 22:30:00 2018-08-10 22:30:00
#> 4 2018-08-10 21:52:00  2018-08-10 22:34:00 2018-08-10 22:34:00
#>         Status     ReisDeel
#> 1    VERTRAAGD c("NS", ....
#> 2 VOLGENS-PLAN NS, Inte....
#> 3 VOLGENS-PLAN c("NS", ....
#> 4 VOLGENS-PLAN NS, Inte....
```

# FAQ

1.  What Can I do with the package?

<!-- end list -->

  - You can access the departures from a station, disruptions (planned
    and unplanned), get travel advise (between stations) and a list of
    all stations (in the Netherlands and some outside.). In the
    [vignette](articles/basic_use_nsapi_package.html) I’ve described how
    to use the functions.

<!-- end list -->

2.  I’m getting a curl timeout\!

<!-- end list -->

  - Yes… That happens, sometimes. The NS website does not always return
    errors but just times you out once in a while. Take a deep breath,
    retry.

<!-- end list -->

3.  Error Bad request or HTTP 400

<!-- end list -->

  - Your username and password might not be set properly

<!-- end list -->

4.  How do I set a password and username?

<!-- end list -->

  - See Authentication in the vignette

## Metadata

The package is MIT licensed although the information from NS is probably
proprietary

``` r
codecoverage <- covr::package_coverage(path = ".",type = "tests")
print(codecoverage)
#> nsapi Coverage: 18.87%
#> R/api_calls_authentification.R: 4.84%
#> R/utils.R: 24.67%
```


NEWS.md
# nsapi 0.1.0.9000
* dev versions


# nsapi 0.1.0

* renamed all functions to start with get_*
* Time is now safely parsed as "Europe/Amsterdam" timezone

# nsapi 0.0.0.9000

* basic version ready
* contains api call, secrets
* all endpoints covered
* pkgdown complete
* vignette with basic functions covered
* sick logo added
* low package coverage




tests/testthat/test_basic_api.R
context("password and username")

test_that("user and password are set", {
  skip_on_cran()
  expect_false(is.na(Sys.getenv("NSAPIACCOUNT", unset = NA)))
  expect_false(is.na(Sys.getenv("NSAPIPW", unset = NA)))
})

test_that("passwordchecker is telling us something", {
  skip_on_cran()
  expect_message(check_ns_api_keys(), regexp = "username is set")
})

context("utility functions")
test_that("utility functions are working", {
  expect_identical(null_to_na(NULL), NA)
  expect_match(message_part(TRUE), regexp = "is set")
  expect_match(message_part(FALSE), regexp = "is NOT set")


  expect_match(datetime("2012-02-27", "12:51"), "2012-02-27T12:51")
  expect_error(datetime(date = 201354, "26587"))
  expect_error(datetime(date = "201354", 26587))
})

test_that("utility function date is working",{ # Naturally dates are messed up on other platforms.
  expect_equal(parse_time("2012-02-27T12:51:00+0100"), as.POSIXct("2012-02-27 12:51:00", tz = "Europe/Amsterdam"))
})


context("travel_advise")

test_that("arguments of travel_advise are tested correctly", {
  skip_on_cran()
  expect_error(get_travel_advise(fromStation = "Amsterdam Centraal", toStation = "Rotterdam Centraal", departure = NA, yearCard = TRUE), regexp = "departure and yearCard can only be TRUE or FALSE")
  expect_error(get_travel_advise(fromStation = "Amsterdam Centraal", toStation = "Rotterdam Centraal", departure = TRUE, yearCard = 1), regexp = "departure and yearCard can only be TRUE or FALSE")
  expect_error(get_travel_advise(fromStation = "Amsterdam Centraal", toStation = "Rotterdam Centraal", departure = TRUE, yearCard = TRUE, previousAdvises = "alpha"), regexp = "need to be numeric")
  expect_error(get_travel_advise(fromStation = "Amsterdam Centraal", toStation = "Rotterdam Centraal", departure = TRUE, yearCard = TRUE, previousAdvises = c(1, 2, 3)), regexp = "need to be numeric")
  expect_error(get_travel_advise(toStation = "Rotterdam Centraal", departure = TRUE, yearCard = TRUE, previousAdvises = 3))
  expect_error(get_travel_advise(toStation = "Rotterdam Centraal", departure = TRUE, yearCard = TRUE, previousAdvises = 3))
  expect_error(get_travel_advise(toStation = NULL, departure = TRUE, yearCard = TRUE, previousAdvises = 1, nextAdvises = 1))
})

test_that("travel_advise returns proper values", {
  skip_on_cran()
  advise_today <- get_travel_advise(fromStation = "Amsterdam Centraal", toStation = "Rotterdam Centraal", departure = TRUE, yearCard = TRUE, previousAdvises = 1, nextAdvises = 1)
  expect_true(is.data.frame(advise_today))
})


# all parsing files.

# empty call
# disruptions_and_maintenance()
context("disruptions")
storing_zonder_ongepland <- list(
  Ongepland = list(
    "\n\t\t\n\t"
  ),
  Gepland = list(
    Storing = list(
      id = list(
        "idnumber1"
      ),
      Traject = list(
        "Station tot station"
      ),
      Periode = list(
        "vector of time"
      ),
      Bericht = list(
        "heel veel tekst"
      )
    ),
    Storing = list(
      id = list(
        "idnumber2"
      ),
      Traject = list(
        "Station tot station"
      ),
      Periode = list(
        "vector of time"
      ),
      Bericht = list(
        "heel veel tekst"
      )
    )
  )
)

storing_met_ongepland <- list(
  Ongepland = list(
    Storing = list(
      id = list("prio-13345"),
      Traject = list("'s-Hertogenbosch-Nijmegen"),
      Reden = list("beperkingen op last van de politie"),
      Datum = list("2010-12-16T11:16:00+0100")
    )
  ),
  Gepland = list(
    "\n\t\t\n\t"
  )
)

test_that("parsing of disruptions works", {
  # checking if testobjects are correct
  expect_true(is.list(storing_zonder_ongepland))
  expect_true(any(names(storing_zonder_ongepland) == c("Ongepland", "Gepland")))
  expect_true(any(names(storing_met_ongepland) == c("Ongepland", "Gepland")))
  testframe1 <- parse_disruptions(storing_zonder_ongepland)
  testframe2 <- parse_disruptions(storing_met_ongepland)
  expect_true(is.data.frame(testframe1))
  expect_true(is.data.frame(testframe2))
  expect_true(all(names(testframe1) == c("id", "Traject", "Periode", "Reden", "Bericht", "Advies", "Datum")))
  expect_true(all(names(testframe2) == c("id", "Traject", "Periode", "Reden", "Bericht", "Advies", "Datum")))
  expect_match(testframe1$id[[1]], "idnumber1")
  expect_match(testframe2$id[[1]], "prio-13345")
  rm(testframe2, testframe1)
})

test_that("disruptions call functions",{
  skip_on_cran()
  expect_error(disruptions_and_maintenance(actual = FALSE,unplanned = FALSE))
  expect_match(class(get_current_disruptions()),"data.frame")
  expect_match(class(get_scheduled_engineering_work()),"data.frame")
  expect_match(class(get_disruptions_station("Utrecht Centraal")),"data.frame")
})


rm(storing_zonder_ongepland)




# parse_reismogelijkheden()

# find_reisdelen()


R/utils.R
# utils
null_to_na <- function(value) {
  ifelse(is.null(value), NA, value)
}

message_part <- function(thing) {
  ifelse(thing, " is set", " is NOT set")
}

# whatever the API documentation says, they return the time in
# timezone Europe/Amsterdam with a offset from UTC.
# That means we can ignore the offset and parse with timezone Europe/Amsterdam.
parse_time <- function(value) {
  as.POSIXct(value, format = "%Y-%m-%dT%H:%M:%S",tz="Europe/Amsterdam")
}



parse_reismogelijkheden <- function(reizen) {
  # pre allocate dataframe
  l_df <- length(reizen)
  holdingframe <- data.frame(
    Melding = I(as.list(rep(NA, l_df))), # this is insane
    AantalOverstappen = integer(l_df),
    GeplandeReisTijd = character(l_df),
    ActueleReisTijd = character(l_df),
    VertrekVertraging = character(l_df),
    AankomstVertraging = character(l_df),
    Optimaal = logical(l_df),
    GeplandeVertrekTijd = as.POSIXct(NA),
    ActueleVertrekTijd = as.POSIXct(NA),
    GeplandeAankomstTijd = as.POSIXct(NA),
    ActueleAankomstTijd = as.POSIXct(NA),
    Status = character(l_df),
    ReisDeel = I(as.list(rep(NA, l_df))),
    stringsAsFactors = FALSE
  )
  # fill the frame
  for (i in seq_len(l_df)) {
    holdingframe$Melding[[i]] <- null_to_na(reizen[[i]]$Melding[[1]])
    holdingframe$AantalOverstappen[[i]] <- as.integer(reizen[[i]]$AantalOverstappen[[1]])
    holdingframe$GeplandeReisTijd[[i]] <- reizen[[i]]$GeplandeReisTijd[[1]]
    holdingframe$ActueleReisTijd[[i]] <- reizen[[i]]$ActueleReisTijd[[1]]
    holdingframe$VertrekVertraging[[i]] <- null_to_na(reizen[[i]]$VertrekVertraging[[1]])
    holdingframe$AankomstVertraging[[i]] <- null_to_na(reizen[[i]]$AankomstVertraging[[1]])
    holdingframe$Optimaal[[i]] <- ifelse(reizen[[i]]$Optimaal[[1]] == "false", FALSE, TRUE)
    holdingframe$GeplandeVertrekTijd[[i]] <- parse_time(reizen[[i]]$GeplandeVertrekTijd[[1]])
    holdingframe$ActueleVertrekTijd[[i]] <- parse_time(reizen[[i]]$ActueleVertrekTijd[[1]])
    holdingframe$GeplandeAankomstTijd[[i]] <- parse_time(reizen[[i]]$GeplandeAankomstTijd[[1]])
    holdingframe$ActueleAankomstTijd[[i]] <- parse_time(reizen[[i]]$ActueleAankomstTijd[[1]])
    holdingframe$Status[[i]] <- reizen[[i]]$Status[[1]]
    holdingframe$ReisDeel[[i]] <- find_reisdelen(reizen[[i]])
  }
  # return the frame
  holdingframe
}



find_reisdelen <- function(reisdeel) {
  ids_reisdeel <- grep("ReisDeel", names(reisdeel))
  l_reisdelen <- length(ids_reisdeel)
  reisdelen <- data.frame(
    Vervoerder = character(l_reisdelen),
    VervoerType = character(l_reisdelen),
    RitNummer = character(l_reisdelen),
    Status = character(l_reisdelen),
    stringsAsFactors = FALSE
  )

  for (i in seq_along(ids_reisdeel)) { # stupid double loop
    index <- ids_reisdeel[[i]]
    # put the top element in
    reisdelen$Vervoerder[[i]] <- reisdeel[[index]]$Vervoerder[[1]]
    reisdelen$VervoerType[[i]] <- reisdeel[[index]]$VervoerType[[1]]
    reisdelen$RitNummer[[i]] <- reisdeel[[index]]$RitNummer[[1]]
    reisdelen$Status[[i]] <- reisdeel[[index]]$Status[[1]]
    # find the ritten
    ids_ritten <- grep("ReisStop", names(reisdeel[[index]]))
    l_ritten <- length(ids_ritten)
    # preallocate dataframe
    rittenframe <- data.frame(
      Naam = character(l_ritten),
      Tijd = as.POSIXlt(NA),
      Spoor = character(l_ritten),
      SpoorWijziging = character(l_ritten),
      stringsAsFactors = FALSE
    )

    for (index_ritten in seq_along(ids_ritten)) {
      j <- ids_ritten[[index_ritten]]
      rittenframe$Naam[[index_ritten]] <- reisdeel[[index]][[j]]$Naam[[1]]
      rittenframe$Tijd[[index_ritten]] <- parse_time(reisdeel[[index]][[j]]$Tijd[[1]])
      rittenframe$Spoor[[index_ritten]] <- null_to_na(reisdeel[[index]][[j]]$Spoor[[1]])
      rittenframe$SpoorWijziging[[index_ritten]] <- null_to_na(attr(reisdeel[[index]][[j]]$Spoor[[1]], "wijziging"))
    }
    reisdelen$Stops[[i]] <- rittenframe
  }
  reisdelen
}


parse_stations <- function(stationslijst) {
  l_stations <- length(stationslijst)
  stationframe <- data.frame(
    Code = character(l_stations),
    Type = character(l_stations),
    Namen = I(as.list(rep(NA, l_stations))),
    Land = character(l_stations),
    UICCode = character(l_stations),
    Lat = character(l_stations),
    Lon = character(l_stations),
    Synoniemen = I(as.list(rep(NA, l_stations))),
    stringsAsFactors = FALSE
  )
  for (i in seq_len(l_stations)) {
    stationframe$Code[[i]] <- stationslijst[[i]]$Code[[1]]
    stationframe$Type[[i]] <- stationslijst[[i]]$Type[[1]]
    stationframe$Namen[[i]] <- stationslijst[[i]]$Namen
    stationframe$Land[[i]] <- stationslijst[[i]]$Land[[1]]
    stationframe$UICCode[[i]] <- stationslijst[[i]]$UICCode[[1]]
    stationframe$Lat[[i]] <- stationslijst[[i]]$Lat[[1]]
    stationframe$Lon[[i]] <- stationslijst[[i]]$Lon[[1]]
    stationframe$Synoniemen[[i]] <- ifelse(unlist(stationslijst[[i]]$Synoniemen) == "\n\t\t\t\n\t\t", NA, stationslijst[[i]]$Synoniemen)
  }
  stationframe$Lat <- as.numeric(stationframe$Lat)
  stationframe$Lon <- as.numeric(stationframe$Lon)
  stationframe
}



parse_vertrekkende_treinen <- function(treinen) {
  l_treinen <- length(treinen)
  treinen_frame <- data.frame(
    RitNummer = character(l_treinen),
    VertrekTijd = as.POSIXct(NA),
    EindBestemming = character(l_treinen),
    TreinSoort = character(l_treinen),
    RouteTekst = character(l_treinen),
    Vervoerder = character(l_treinen),
    VertrekSpoor = character(l_treinen),
    Spoorwijziging = character(l_treinen),
    stringsAsFactors = FALSE
  )
  for (i in seq_len(l_treinen)) {
    treinen_frame$Ritnummer[[i]] <- treinen[[i]]$Ritnummer[[1]]
    treinen_frame$VertrekTijd[[i]] <- parse_time(treinen[[i]]$VertrekTijd[[1]])
    treinen_frame$EindBestemming[[i]] <- treinen[[i]]$EindBestemming[[1]]
    treinen_frame$TreinSoort[[i]] <- treinen[[i]]$TreinSoort[[1]]
    treinen_frame$RouteTekst[[i]] <- null_to_na(treinen[[i]]$RouteTekst[[1]])
    treinen_frame$VertrekSpoor[[i]] <- treinen[[i]]$VertrekSpoor[[1]]
    treinen_frame$SpoorWijziging[[i]] <- attr(treinen[[i]]$SpoorWijziging, "wijziging")
  }
  treinen_frame
}



deal_with_response <- function(response) {
  response$raise_for_status()
  list_response <- xml2::as_list(xml2::as_xml_document(response$parse("UTF-8")))[[1]]
  list_response
}



parse_disruptions <- function(disruptions) {
  l_gepl <- length(disruptions$Gepland)
  if (l_gepl == 1L) {
    if (any(disruptions$Gepland[[1]] == "\n\t\t\n\t")) l_gepl <- 0L
  }

  l_ongpl <- length(disruptions$Ongepland)
  if (l_ongpl == 1L) {
    if (any(disruptions$Ongepland[[1]] == "\n\t\t\n\t")) l_ongpl <- 0L
  }

  tot <- l_gepl + l_ongpl
  if(tot == 0){
    stop("The API returned empty results\nTry changing actual and unplanned")
  }

  holdingframe <- data.frame(
    id = character(tot),
    Traject = character(tot),
    Periode = character(tot),
    Reden = character(tot),
    Bericht = character(tot),
    Advies = character(tot),
    Datum = as.POSIXct(NA),
    stringsAsFactors = FALSE
  )
  for (i in seq_len(l_gepl)) {
    holdingframe$id[[i]] <- disruptions$Gepland[[i]]$id[[1]]
    holdingframe$Traject[[i]] <- disruptions$Gepland[[i]]$Traject[[1]]
    holdingframe$Periode[[i]] <- null_to_na(disruptions$Gepland[[i]]$Periode[[1]])
    holdingframe$Bericht[[i]] <- null_to_na(disruptions$Gepland[[i]]$Bericht[[1]])
    holdingframe$Advies[[i]] <- null_to_na(disruptions$Gepland[[i]]$Advies[[1]])
    holdingframe$Datum[[i]] <- parse_time(null_to_na(disruptions$Gepland[[i]]$Datum[[1]]))
  }

  for (j in seq_len(l_ongpl)) {
    index <- l_gepl + j
    holdingframe$id[[index]] <- disruptions$Ongepland[[j]]$id[[1]]
    holdingframe$Traject[[index]] <- disruptions$Ongepland[[j]]$Traject[[1]]
    holdingframe$Periode[[index]] <- null_to_na(disruptions$Ongepland[[j]]$Periode[[1]])
    holdingframe$Bericht[[index]] <- null_to_na(disruptions$Ongepland[[j]]$Bericht[[1]])
    holdingframe$Advies[[index]] <- null_to_na(disruptions$Ongepland[[j]]$Advies[[1]])
    holdingframe$Datum[[index]] <- parse_time(null_to_na(disruptions$Ongepland[[j]]$Datum[[1]]))
  }
  holdingframe
}


R/nsapi-package.R
#' NSapi package
#'
#' Access the NS api and download current departure times,
#' disruptions and engineering work, the station list, and
#' travel recommendations from station to station. All results will be
#' returned as a 'data.frame'.
#' NS (Nederlandse Spoorwegen; Dutch Railways) is the largest train travel
#' provider in the Netherlands. for more information about the API itself
#' see <https://www.ns.nl/en/travel-information/ns-api>.
#' To use the API, and this package, you will need to obtain a username
#' and password. More information about authentication and the use of the functions
#' are described in the vignette.
#'
#' Implemented functions:
#'
#' - Get travel advise between two stations \code{\link{get_travel_advise}}.
#' - Get departures from a particular station  \code{\link{get_departures}}.
#' - All the stations \code{\link{get_stationlist}}.
#' - Get all current disruptions \code{\link{get_current_disruptions}}.
#' - Get all scheduled engineering in next two weeks \code{\link{get_scheduled_engineering_work}}.
#' - Get all (un-)scheduled disruptions \code{\link{get_disruptions_station}}.
#' - Find out if you have configured the API keys correct \code{\link{check_ns_api_keys}}.
#' - A small helper function to create datetimestamp \code{\link{datetime}}.
#'
#'
#' Image use:
#' I have combined several images from the Noun project:
#'
#' - hands on circle: 		Data by Gregor Cresnar from the Noun Project:
#' - train: 				railway by BomSymbols from the Noun Project
#' - laptop				Laptop by abdul karim from the Noun Project
#'
#' @docType package
#' @name nsapi
NULL



R/api_calls_authentification.R
# basic functionality
# key management and authentication

#' Are the username and password set?
#'
#' A function to see if your username and password are set for nsapi.
#' Note that I have no clue if the password or username are correct,
#' that is up to you.
#'
#' For more details see the help vignette:
#' \code{vignette("basic_use_nsapi_package", package = "nsapi")}
#' and get your keys here:
#' \url{https://www.ns.nl/ews-aanvraagformulier/?0}
#'
#' @export
check_ns_api_keys <- function() {
  pw <- !is.na(Sys.getenv("NSAPIPW", unset = NA))
  username <- !is.na(Sys.getenv("NSAPIACCOUNT", unset = NA))
  message(paste0(
    "Your username", message_part(pw),
    ",\nyour password", message_part(username),
    ifelse(all(pw, username), "",
           "\nSee the vignette `Basic use of the nsapi package` section: 'Authentication'\n for more information on how to set these variables")
  ))
}



# call api


nsapi_client <- function(){
  crul::HttpClient$new(
  url = "webservices.ns.nl",
  auth = crul::auth(
    user = Sys.getenv("NSAPIACCOUNT", unset = NA),
    pwd = Sys.getenv("NSAPIPW", unset = NA),
    auth = "basic"
  ),
  opts = list(timeout_ms = 3000),
  headers = list(`User-Agent` = "ns_api_r_package")
)
}

#' Translate date and time into datetime stamp to use in api call
#'
#' The datetime stamp format might be a bit difficult to use. This
#' helper function allows you to supply the date and time and a datetime
#' character string comes out.
#'
#' For example if it is 21-08-2018, 15.21 hour in the Netherlands than you can
#' add these two items into the function.
#'
#' @param date date in iso format (year-month-day): 2018-12-29 for example
#' @param time time in standard format (HH:MM): 20:21
#' @export
#' @examples
#' \dontrun{
#' get_travel_advise(
#'     fromStation = "Amsterdam Centraal",
#'     toStation = "Utrecht Centraal",
#'     dateTime = datetime("2018-08-21","15:21"),
#'     departure = TRUE
#'     )
#' }
datetime <- function(date, time) {
  if (!is.character(date)) stop(paste(date, " is not a character value, \ndate needs to be YYYY-MM-DD format"))
  if (!is.character(time)) stop(paste(time, " is not a character value, \ntime needs to be in HH:MM format"))
  paste0(date, "T", time)
}

#' Get travel advise from one station to another station
#'
#' This is equivalent to the [NS reisplanner](https://www.ns.nl/reisplanner#/),
#' you give in a from and to station, the timestamp, if you want the time
#' to be your departure or arrival time, and optionally if you have a NS year
#' card (has effect on some travels).
#'
#' You can also specify how many trips before and after the chosen time you
#' want to collect (defaults to 4, maximum is 5 before, and 5 after).
#'
#' Some things to consider: station names need to be in Dutch but the NS
#' webservice also accepts shortened versions: "Utrecht Centraal" and
#' "ut" is apparently the same.
#' Station names can be found with the `get_stationlist()` call.
#'
#' Although the documentation
#' <https://www.ns.nl/en/travel-information/ns-api/documentation-travel-recommendations.html>
#' is in English, the returned values are all in Dutch. And I keep the results
#' in Dutch.
#'
#'
#' @param fromStation the station to start from, for instance "Rotterdam Centraal"
#' @param toStation the station to end, for instance "Utrecht Centraal"
#' @param dateTime defaults to current time, but you can use a different one: f.i. 2012-02-21T15:50, You can also use the `datetime()` function.
#' @param departure is the datetime the start or end time? do you want to depart on that date or arrive, defaults to departure
#' @param yearCard if you have a NS year card (jaarabonnement) some trips will be different
#' @param hslAllowed use of the high speed train
#' @param previousAdvises how many advices do you want before the time
#' @param nextAdvises how many advises do you want after
#' @return A dataframe with travel advises around your chosen date time. Date time columns are formatted as time in "Europe/Amsterdam" timezone.
#' @export
#' @examples
#' \dontrun{
#' get_travel_advise("Amsterdam Centraal",
#' "Utrecht Centraal",dateTime = "2018-08-01T15:21",departure = TRUE)
#' }
get_travel_advise <- function(fromStation, toStation, dateTime = NULL,
                          departure = TRUE, yearCard = FALSE,
                          hslAllowed = FALSE,
                          previousAdvises = 4, nextAdvises = 4) {
  if (any(base::missing(fromStation), base::missing(toStation), is.null(fromStation), is.null(toStation))) stop("You need to supply both fromStation and toStation")
  if (!all(is.logical(departure), is.logical(yearCard), !is.na(departure), !is.na(yearCard))) stop("departure and yearCard can only be TRUE or FALSE")
  if (!all(is.numeric(previousAdvises), is.numeric(nextAdvises), length(previousAdvises) == 1, length(nextAdvises) == 1)) stop("previousAdvises and nextAdvises need to be numeric and a single number, f.i. 8.")

  nsapi_client <- nsapi_client()
  response <- nsapi_client$get(
    path = "/ns-api-treinplanner",
    query = list(
      fromStation = fromStation,
      toStation = toStation,
      departure = departure,
      dateTime = dateTime,
      yearCard = yearCard,
      hslAllowed = hslAllowed,
      previousAdvices = previousAdvises,
      nextAdvices = nextAdvises
    )
  )

  list_response <- deal_with_response(response)
  parse_reismogelijkheden(list_response)
}

#' Get a complete list of all the stations
#'
#' This function should not be called too often.
#' The stations will probably not change a lot so it might
#' be better to save it as a dataframe in your local environment for further use.
#'
#' The dataframe consists of all the Dutch stations, many German and Belgian stations and
#' bigger stations in other countries in Europe.
#' \url{https://www.ns.nl/en/travel-information/ns-api/documentation-station-list.html}
#' @export
get_stationlist <- function() {
  nsapi_client <- nsapi_client()
  response <- nsapi_client$get(path = "/ns-api-stations-v2")
  list_response <- deal_with_response(response)
  parse_stations(list_response)
}

#' Get up to date departures from a station
#'
#' Shows up to date departure times for a station. Displays all departing
#' trains for the next hour.
#' At least 10 departure times will be sent back and at least all the
#' departure times for the next hour.
#' \url{https://www.ns.nl/en/travel-information/ns-api/documentation-up-to-date-departure-times.html}.
#' @param station station names need to be in Dutch and the NS webservice also accepts short versions:f.i. Groningen or GN
#' @return a dataframe with date time formatted as time in "Europe/Amsterdam" timezone.
#' @export
#' @return a dataframe with departure times at this moment on the station you chose. Date time columns are formatted as time in "Europe/Amsterdam" timezone.
#' @examples
#' \dontrun{
#' get_departures("UT")
#' }
get_departures <- function(station) {
  nsapi_client <- nsapi_client()
  response <- nsapi_client$get(path = "/ns-api-avt", query = list(station = station))
  list_response <- deal_with_response(response)
  parse_vertrekkende_treinen(list_response)
}







disruptions_and_maintenance <- function(station = NULL, actual = NULL, unplanned = NULL) {
  nsapi_client <- nsapi_client()
  response <- nsapi_client$get(
    path = "/ns-api-storingen",
    query = list(
      station = station,
      actual = actual,
      unplanned = unplanned
    )
  )
  list_response <- deal_with_response(response)
  parse_disruptions(list_response)
}


#' Get disruptions and engineering work
#'
#' These 3 functions call out to find out about disruptions and engineering on
#' the tracks for the current time, for the next 2 weeks or a specific station.
#'
#' @details
#' - current disruptions  (=unscheduled disruptions + current engineering work)
#'
#' These are all the disruptions of the railroad at this moment. So both
#' unscheduled work and work
#' and work that was scheduled and currently underway.
#' Use `get_current_disruptions()`
#'
#' - scheduled engineering work(=scheduled engineering work)
#'
#' Get all the scheduled engineering work for the next 2 weeks
#' with `get_scheduled_engineering_work()`. This will exclude
#' work that was unplanned.
#'
#' - current disruptions for a specific station (=unscheduled disruptions + current engineering work)
#'
#' Use `get_disruptions_station()` and give a station name as argument.
#'
#' \url{https://www.ns.nl/en/travel-information/ns-api/documentation-disruptions-and-maintenance-work.html}
#' @param station  optional, station names need to be in Dutch and the NS webservice also accepts short versions:f.i. Groningen or GN
#' @return a dataframe with disruptions. Date time columns are formatted as time in "Europe/Amsterdam" timezone.
#' @export
get_disruptions_station <- function(station){
  disruptions_and_maintenance(station= station)
}

#' @describeIn get_disruptions_station current disruptions
#' @export
get_current_disruptions <- function(){
  disruptions_and_maintenance(station = NULL, actual = TRUE,unplanned = NULL)
}

#' @describeIn get_disruptions_station scheduled disruptions
#' @export
get_scheduled_engineering_work <- function(){
  disruptions_and_maintenance(station = NULL,actual = FALSE, unplanned = TRUE)
}
